Fix cloudfront_distribution disabling ipv6 when updating resource

abikouo opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
IpV6Enabled value is changed when updating a cloud distribution
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
#336 
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
When option ipv6_enabled is not explicitly specified as parameters, the current value is kept when updating the cloudfront distribution
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

tremble commented on plugins/modules/cloudfront_distribution.py:
```diff
@@ -1975,7 +1975,7 @@ def validate_restrictions(self, config, restrictions, purge_restrictions=False):
         try:
             config['default_root_object'] = default_root_object or config.get('default_root_object', '')
-            config['is_i_p_v_6_enabled'] = ipv6_enabled or config.get('i_p_v_6_enabled', self.__default_ipv6_enabled)
+            config['is_i_p_v6_enabled'] = ipv6_enabled if ipv6_enabled is not None else config.get('is_i_p_v6_enabled', self.__default_ipv6_enabled)
```
The current logic looks like disabling IPv6 wouldn't be possible, so this is probably an improvement.  

We really need some integration tests...

abikouo replied:
> I am unsure about this change. I think the main problem is due to the fact that if the user doesn't explicitly specify in every task `ipv6_enabled=True` it is overwritten by the default value (`False`). I would prefer to have your opinion @tremble @jillr.

you should not change the value of ``ipv6_enable`` when updating a cloudfront without this option, the value should remain as it was before updating, and here it is not the case, I will introduce some test to highlight

abikouo commented on plugins/modules/cloudfront_distribution.py:
```diff
@@ -2056,7 +2056,7 @@ def validate_distribution_from_aliases_caller_reference(self, distribution_id, a
                 return self.validate_distribution_from_caller_reference(caller_reference)
             else:
-                if aliases:
+                if aliases and distribution_id is None:
```
``alias`` and ``distribution_id`` are mutually exclusive, it is not set for ``aliases`` but I think it was intended. 
we can skip this yes

abikouo commented:
@tremble @alinabuzachis @jillr 

Integrated tests have been updated for the cloudfront_distribution module 
here after the trace before the fix 

```
TASK [cloudfront_distribution : create cloudfront distribution using defaults] **********************************************************************************************
changed: [localhost] => {"active_trusted_key_groups": {"enabled": false, "quantity": 0}, "active_trusted_signers": {"enabled": false, "quantity": 0}, "aliases": {"quantity": 0}, "arn": "arn:aws:cloudfront::721066863947:distribution/E1AX836JQA0NZS", "cache_behaviors": {"quantity": 0}, "caller_reference": "2021-04-14T18:11:11.157258", "changed": true, "comment": "Distribution created by Ansible with datetime stamp 2021-04-14T18:11:11.157258", "custom_error_responses": {"quantity": 0}, "default_cache_behavior": {"allowed_methods": {"cached_methods": {"items": ["HEAD", "GET"], "quantity": 2}, "items": ["HEAD", "GET"], "quantity": 2}, "compress": false, "default_ttl": 86400, "field_level_encryption_id": "", "forwarded_values": {"cookies": {"forward": "none"}, "headers": {"quantity": 0}, "query_string": true, "query_string_cache_keys": {"quantity": 0}}, "lambda_function_associations": {"quantity": 0}, "max_ttl": 31536000, "min_ttl": 0, "smooth_streaming": false, "target_origin_id": "cloudteam01-origin.example.com", "trusted_key_groups": {"enabled": false, "quantity": 0}, "trusted_signers": {"enabled": false, "quantity": 0}, "viewer_protocol_policy": "allow-all"}, "default_root_object": "", "domain_name": "d3boyln1vh4h88.cloudfront.net", "enabled": true, "http_version": "http2", "id": "E1AX836JQA0NZS", "in_progress_invalidation_batches": 0, "is_ipv6_enabled": false, "last_modified_time": "2021-04-14T16:11:11.710000+00:00", "logging": {"bucket": "", "enabled": false, "include_cookies": false, "prefix": ""}, "origin_groups": {"quantity": 0}, "origins": {"items": [{"connection_attempts": 3, "connection_timeout": 10, "custom_headers": {"quantity": 0}, "custom_origin_config": {"http_port": 80, "https_port": 443, "origin_keepalive_timeout": 5, "origin_protocol_policy": "match-viewer", "origin_read_timeout": 30, "origin_ssl_protocols": {"items": ["TLSv1", "TLSv1.1", "TLSv1.2"], "quantity": 3}}, "domain_name": "cloudteam01-origin.example.com", "id": "cloudteam01-origin.example.com", "origin_path": "", "origin_shield": {"enabled": false}}], "quantity": 1}, "price_class": "PriceClass_All", "restrictions": {"geo_restriction": {"quantity": 0, "restriction_type": "none"}}, "status": "InProgress", "tags": {}, "viewer_certificate": {"certificate_source": "cloudfront", "cloud_front_default_certificate": true, "minimum_protocol_version": "TLSv1"}, "web_acl_id": ""}

TASK [cloudfront_distribution : set_fact] ***********************************************************************************************************************************
ok: [localhost] => {"ansible_facts": {"distribution_id": "E1AX836JQA0NZS"}, "changed": false}

TASK [cloudfront_distribution : ensure that default value of ipv6_enabled] **************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [cloudfront_distribution : Update the distribution with tags] **********************************************************************************************************
changed: [localhost] => {"active_trusted_key_groups": {"enabled": false, "quantity": 0}, "active_trusted_signers": {"enabled": false, "quantity": 0}, "aliases": {"quantity": 0}, "arn": "arn:aws:cloudfront::721066863947:distribution/E1AX836JQA0NZS", "cache_behaviors": {"quantity": 0}, "caller_reference": "2021-04-14T18:11:11.157258", "changed": true, "comment": "Distribution created by Ansible with datetime stamp 2021-04-14T18:11:11.157258", "custom_error_responses": {"quantity": 0}, "default_cache_behavior": {"allowed_methods": {"cached_methods": {"items": ["HEAD", "GET"], "quantity": 2}, "items": ["HEAD", "GET"], "quantity": 2}, "compress": false, "default_ttl": 86400, "field_level_encryption_id": "", "forwarded_values": {"cookies": {"forward": "none"}, "headers": {"quantity": 0}, "query_string": true, "query_string_cache_keys": {"quantity": 0}}, "lambda_function_associations": {"quantity": 0}, "max_ttl": 31536000, "min_ttl": 0, "smooth_streaming": false, "target_origin_id": "cloudteam01-origin.example.com", "trusted_key_groups": {"enabled": false, "quantity": 0}, "trusted_signers": {"enabled": false, "quantity": 0}, "viewer_protocol_policy": "allow-all"}, "default_root_object": "", "domain_name": "d3boyln1vh4h88.cloudfront.net", "enabled": true, "http_version": "http2", "id": "E1AX836JQA0NZS", "in_progress_invalidation_batches": 0, "is_ipv6_enabled": false, "last_modified_time": "2021-04-14T16:11:14.213000+00:00", "logging": {"bucket": "", "enabled": false, "include_cookies": false, "prefix": ""}, "origin_groups": {"quantity": 0}, "origins": {"items": [{"connection_attempts": 3, "connection_timeout": 10, "custom_headers": {"quantity": 0}, "custom_origin_config": {"http_port": 80, "https_port": 443, "origin_keepalive_timeout": 5, "origin_protocol_policy": "match-viewer", "origin_read_timeout": 30, "origin_ssl_protocols": {"items": ["TLSv1", "TLSv1.1", "TLSv1.2"], "quantity": 3}}, "domain_name": "cloudteam01-origin.example.com", "id": "cloudteam01-origin.example.com", "origin_path": "", "origin_shield": {"enabled": false}}], "quantity": 1}, "price_class": "PriceClass_All", "restrictions": {"geo_restriction": {"quantity": 0, "restriction_type": "none"}}, "status": "InProgress", "tags": {"property": "ansible"}, "viewer_certificate": {"certificate_source": "cloudfront", "cloud_front_default_certificate": true, "minimum_protocol_version": "TLSv1"}, "web_acl_id": ""}

TASK [cloudfront_distribution : ensure the 'ipv6_enabled' value has not changed] ********************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [cloudfront_distribution : Update the distribution ipv6_enabled set to true] *******************************************************************************************
changed: [localhost] => {"active_trusted_key_groups": {"enabled": false, "quantity": 0}, "active_trusted_signers": {"enabled": false, "quantity": 0}, "aliases": {"quantity": 0}, "arn": "arn:aws:cloudfront::721066863947:distribution/E1AX836JQA0NZS", "cache_behaviors": {"quantity": 0}, "caller_reference": "2021-04-14T18:11:11.157258", "changed": true, "comment": "Distribution created by Ansible with datetime stamp 2021-04-14T18:11:11.157258", "custom_error_responses": {"quantity": 0}, "default_cache_behavior": {"allowed_methods": {"cached_methods": {"items": ["HEAD", "GET"], "quantity": 2}, "items": ["HEAD", "GET"], "quantity": 2}, "compress": false, "default_ttl": 86400, "field_level_encryption_id": "", "forwarded_values": {"cookies": {"forward": "none"}, "headers": {"quantity": 0}, "query_string": true, "query_string_cache_keys": {"quantity": 0}}, "lambda_function_associations": {"quantity": 0}, "max_ttl": 31536000, "min_ttl": 0, "smooth_streaming": false, "target_origin_id": "cloudteam01-origin.example.com", "trusted_key_groups": {"enabled": false, "quantity": 0}, "trusted_signers": {"enabled": false, "quantity": 0}, "viewer_protocol_policy": "allow-all"}, "default_root_object": "", "domain_name": "d3boyln1vh4h88.cloudfront.net", "enabled": true, "http_version": "http2", "id": "E1AX836JQA0NZS", "in_progress_invalidation_batches": 0, "is_ipv6_enabled": true, "last_modified_time": "2021-04-14T16:11:16.457000+00:00", "logging": {"bucket": "", "enabled": false, "include_cookies": false, "prefix": ""}, "origin_groups": {"quantity": 0}, "origins": {"items": [{"connection_attempts": 3, "connection_timeout": 10, "custom_headers": {"quantity": 0}, "custom_origin_config": {"http_port": 80, "https_port": 443, "origin_keepalive_timeout": 5, "origin_protocol_policy": "match-viewer", "origin_read_timeout": 30, "origin_ssl_protocols": {"items": ["TLSv1", "TLSv1.1", "TLSv1.2"], "quantity": 3}}, "domain_name": "cloudteam01-origin.example.com", "id": "cloudteam01-origin.example.com", "origin_path": "", "origin_shield": {"enabled": false}}], "quantity": 1}, "price_class": "PriceClass_All", "restrictions": {"geo_restriction": {"quantity": 0, "restriction_type": "none"}}, "status": "InProgress", "tags": {"property": "ansible"}, "viewer_certificate": {"certificate_source": "cloudfront", "cloud_front_default_certificate": true, "minimum_protocol_version": "TLSv1"}, "web_acl_id": ""}

TASK [cloudfront_distribution : ensure the 'ipv6_enabled' value has changed (new value is true)] ****************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [cloudfront_distribution : Update the distribution with tags] **********************************************************************************************************
changed: [localhost] => {"active_trusted_key_groups": {"enabled": false, "quantity": 0}, "active_trusted_signers": {"enabled": false, "quantity": 0}, "aliases": {"quantity": 0}, "arn": "arn:aws:cloudfront::721066863947:distribution/E1AX836JQA0NZS", "cache_behaviors": {"quantity": 0}, "caller_reference": "2021-04-14T18:11:11.157258", "changed": true, "comment": "Distribution created by Ansible with datetime stamp 2021-04-14T18:11:11.157258", "custom_error_responses": {"quantity": 0}, "default_cache_behavior": {"allowed_methods": {"cached_methods": {"items": ["HEAD", "GET"], "quantity": 2}, "items": ["HEAD", "GET"], "quantity": 2}, "compress": false, "default_ttl": 86400, "field_level_encryption_id": "", "forwarded_values": {"cookies": {"forward": "none"}, "headers": {"quantity": 0}, "query_string": true, "query_string_cache_keys": {"quantity": 0}}, "lambda_function_associations": {"quantity": 0}, "max_ttl": 31536000, "min_ttl": 0, "smooth_streaming": false, "target_origin_id": "cloudteam01-origin.example.com", "trusted_key_groups": {"enabled": false, "quantity": 0}, "trusted_signers": {"enabled": false, "quantity": 0}, "viewer_protocol_policy": "allow-all"}, "default_root_object": "", "domain_name": "d3boyln1vh4h88.cloudfront.net", "enabled": true, "http_version": "http2", "id": "E1AX836JQA0NZS", "in_progress_invalidation_batches": 0, "is_ipv6_enabled": false, "last_modified_time": "2021-04-14T16:11:18.619000+00:00", "logging": {"bucket": "", "enabled": false, "include_cookies": false, "prefix": ""}, "origin_groups": {"quantity": 0}, "origins": {"items": [{"connection_attempts": 3, "connection_timeout": 10, "custom_headers": {"quantity": 0}, "custom_origin_config": {"http_port": 80, "https_port": 443, "origin_keepalive_timeout": 5, "origin_protocol_policy": "match-viewer", "origin_read_timeout": 30, "origin_ssl_protocols": {"items": ["TLSv1", "TLSv1.1", "TLSv1.2"], "quantity": 3}}, "domain_name": "cloudteam01-origin.example.com", "id": "cloudteam01-origin.example.com", "origin_path": "", "origin_shield": {"enabled": false}}], "quantity": 1}, "price_class": "PriceClass_All", "restrictions": {"geo_restriction": {"quantity": 0, "restriction_type": "none"}}, "status": "InProgress", "tags": {"property": "ansible", "test": "integration"}, "viewer_certificate": {"certificate_source": "cloudfront", "cloud_front_default_certificate": true, "minimum_protocol_version": "TLSv1"}, "web_acl_id": ""}

TASK [cloudfront_distribution : ensure the 'ipv6_enabled' value has not changed (value remains true)] ***********************************************************************
fatal: [localhost]: FAILED! => {
    "assertion": "cf_update_tags.is_ipv6_enabled",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}

```

Now with the fix everything is working fine.
